File: cli/src/quality.rs
use regex::Regex;
use std::sync::OnceLock;

static NOISY_PATTERNS: OnceLock<Regex> = OnceLock::new();
static JARGON_PATTERNS: OnceLock<Regex> = OnceLock::new();

// Quality scoring thresholds
const THRESHOLD_NOISE: usize = 6;
const THRESHOLD_JARGON: usize = 3;
const THRESHOLD_MIN_CHARS: usize = 500;

// Quality scoring penalties
const PENALTY_TOO_SHORT: f32 = 0.25;
const PENALTY_MISSING_LINKS: f32 = 0.10;
const PENALTY_DUPLICATE_HEAVY: f32 = 0.15;
const PENALTY_NOISY: f32 = 0.10;
const PENALTY_JARGON: f32 = 0.10;

// Quality scoring bonuses
const BONUS_HAS_FRONTMATTER: f32 = 0.05;
const BONUS_HAS_STRUCTURAL_ANCHORS: f32 = 0.05;

#[derive(Debug, Clone)]
pub struct QualityScore {
    pub score: f32,
    pub too_short: bool,
    pub missing_links: bool,
    pub duplicate_heavy: bool,
    pub noisy: bool,
    pub acceptable: bool,
}

pub fn score_content(markdown: &str, links: &[String], threshold: f32) -> QualityScore {
    let trimmed = markdown.trim();
    let len = trimmed.len();

    let too_short = len < THRESHOLD_MIN_CHARS;
    let missing_links = links.is_empty();

    // Optimize duplicate detection: pre-calculate total lines via byte scan and early exit
    let duplicate_heavy = if trimmed.is_empty() {
        false
    } else {
        let total_lines = trimmed.bytes().filter(|&b| b == b'\n').count() + 1;
        let threshold_unique = std::cmp::max(5, total_lines / 3);
        let mut unique_set = std::collections::HashSet::with_capacity(std::cmp::min(
            total_lines,
            threshold_unique + 1,
        ));
        let mut is_dup = true;
        for line in trimmed.lines() {
            unique_set.insert(line);
            if unique_set.len() >= threshold_unique {
                is_dup = false;
                break;
            }
        }
        is_dup
    };

    // Optimize noise detection: use case-insensitive regex to avoid to_lowercase() allocation
    let noisy_re = NOISY_PATTERNS.get_or_init(|| {
        Regex::new("(?i)cookie|subscribe|javascript|log in|sign up")
            .expect("Invalid quality noise regex patterns")
    });
    // Early exit if noise threshold is exceeded
    let noisy_count = noisy_re
        .find_iter(trimmed)
        .take(THRESHOLD_NOISE + 1)
        .count();
    let noisy = noisy_count > THRESHOLD_NOISE;

    let jargon_re = JARGON_PATTERNS.get_or_init(|| {
        Regex::new("(?i)seamlessly|robust|powerful|comprehensive|streamlined|leverage|revolutionize|game-changing|intuitive|next-generation|cutting-edge|state-of-the-art|best-in-class|unlock|transform|supercharge")
            .expect("Invalid quality jargon regex patterns")
    });
    // Early exit if jargon threshold is exceeded
    let jargon_count = jargon_re
        .find_iter(trimmed)
        .take(THRESHOLD_JARGON + 1)
        .count();
    let jargon_heavy = jargon_count > THRESHOLD_JARGON;

    let has_frontmatter = if let Some(rest) = trimmed.strip_prefix("---") {
        let header_end = rest
            .find("\n---")
            .map(|i| i + 7)
            .unwrap_or(trimmed.len());
        let header_slice = &trimmed[..header_end];
        header_slice.contains("relevance_score:")
            && header_slice.contains("intent_category:")
            && header_slice.contains("token_estimate:")
            && header_slice.contains("last_updated:")
    } else {
        false
    };

    let has_structural_anchors = trimmed.contains("[ANCHOR:")
        && trimmed.contains("[ANCHOR: SUMMARY]")
        && trimmed.contains("[ANCHOR: TECHNICAL_DETAILS]")
        && trimmed.contains("[ANCHOR: COMPARISON]")
        && trimmed.contains("[ANCHOR: CITATIONS]");

    let mut score = 1.0_f32;
    if too_short {
        score -= PENALTY_TOO_SHORT;
    }
    if missing_links {
        score -= PENALTY_MISSING_LINKS;
    }
    if duplicate_heavy {
        score -= PENALTY_DUPLICATE_HEAVY;
    }
    if noisy {
        score -= PENALTY_NOISY;
    }
    if jargon_heavy {
        score -= PENALTY_JARGON;
    }

    if has_frontmatter {
        score += BONUS_HAS_FRONTMATTER;
    }
    if has_structural_anchors {
        score += BONUS_HAS_STRUCTURAL_ANCHORS;
    }

    let score = score.clamp(0.0, 1.0);
    let acceptable = score >= threshold && !too_short;

    QualityScore {
        score,
        too_short,
        missing_links,
        duplicate_heavy,
        noisy: noisy || jargon_heavy,
        acceptable,
    }
}

File: scripts/quality.py
"""
Heuristics for scoring the quality of resolved content.
"""

from dataclasses import dataclass

# Quality scoring penalties
PENALTY_TOO_SHORT = 0.25
PENALTY_MISSING_LINKS = 0.10
PENALTY_DUPLICATE_HEAVY = 0.15
PENALTY_NOISY = 0.10
PENALTY_JARGON = 0.10

# Quality scoring bonuses
BONUS_HAS_FRONTMATTER = 0.05
BONUS_HAS_ANCHORS = 0.05

# Quality scoring thresholds
THRESHOLD_NOISE = 6
THRESHOLD_JARGON = 3
THRESHOLD_MIN_CHARS = 500

# Acceptance threshold
ACCEPTABLE_THRESHOLD = 0.65

# Bot challenge detection signals (case-insensitive substring match)
_BOT_CHALLENGE_SIGNALS: frozenset[str] = frozenset(
    [
        "cf-challenge",  # Cloudflare challenge page meta tag
        "cf_chl_opt",  # Cloudflare JS challenge var
        "ray id",  # Cloudflare Ray ID footer
        "ddos-guard",  # DDoS-Guard service
        "please enable javascript",
        "enable cookies",
        "checking your browser",  # Cloudflare "Checking your browser..."
        "just a moment",  # Cloudflare interstitial title
        "security check",
        "access denied",
        "403 forbidden",
        "are you a human",
        "prove you are human",
        "recaptcha",
        "hcaptcha",
    ]
)


@dataclass
class QualityScore:
    score: float
    too_short: bool
    missing_links: bool
    duplicate_heavy: bool
    noisy: bool
    acceptable: bool
    bot_challenge: bool = False


def _check_duplicates(text: str) -> bool:
    """Check if content has excessive duplicate lines."""
    lines = text.splitlines()
    num_lines = len(lines)
    if num_lines == 0:
        return False
    unique_lines = len({line.strip() for line in lines if line.strip()})
    return unique_lines < max(5, num_lines // 3)


def _check_jargon(text_lower: str) -> bool:
    """Check if content contains excessive marketing jargon/AI slop."""
    jargon_signals = [
        "seamlessly",
        "robust",
        "powerful",
        "comprehensive",
        "streamlined",
        "leverage",
        "revolutionize",
        "game-changing",
        "intuitive",
        "next-generation",
        "cutting-edge",
        "state-of-the-art",
        "best-in-class",
        "unlock",
        "transform",
        "supercharge",
    ]
    jargon_count = sum(text_lower.count(signal) for signal in jargon_signals)
    return jargon_count > THRESHOLD_JARGON


def _check_frontmatter(text: str) -> bool:
    """Check for 2026 standard YAML frontmatter fields."""
    required_yaml = [
        "relevance_score:",
        "intent_category:",
        "token_estimate:",
        "last_updated:",
    ]
    return text.startswith("---") and all(field in text for field in required_yaml)


def _check_anchors(text: str) -> bool:
    """Check for mandatory RAG-optimized structural anchors."""
    required_anchors = [
        "[ANCHOR: SUMMARY]",
        "[ANCHOR: TECHNICAL_DETAILS]",
        "[ANCHOR: COMPARISON]",
        "[ANCHOR: CITATIONS]",
    ]
    return all(anchor in text for anchor in required_anchors)


def _compute_noise(text_lower: str) -> bool:
    """Detect noisy signals like cookie/subscribe prompts."""
    noisy_signals = ["cookie", "subscribe", "javascript", "log in", "sign up"]
    noise_count = sum(text_lower.count(signal) for signal in noisy_signals)
    return noise_count > THRESHOLD_NOISE


def is_bot_challenge(content: str) -> bool:
    """Return True if content looks like a bot-detection interstitial.

    Checks a representative sample of the content (first 2000 chars)
    to avoid scanning megabyte-sized pages.
    """
    sample = content[:2000].lower()
    return any(signal in sample for signal in _BOT_CHALLENGE_SIGNALS)


def _compute_penalties(
    score: float,
    too_short: bool,
    missing_links: bool,
    duplicate_heavy: bool,
    noisy: bool,
    jargon_heavy: bool,
) -> float:
    """Apply quality penalties."""
    if too_short:
        score -= PENALTY_TOO_SHORT
    if missing_links:
        score -= PENALTY_MISSING_LINKS
    if duplicate_heavy:
        score -= PENALTY_DUPLICATE_HEAVY
    if noisy:
        score -= PENALTY_NOISY
    if jargon_heavy:
        score -= PENALTY_JARGON
    return score


def _compute_bonuses(score: float, has_frontmatter: bool, has_anchors: bool) -> float:
    """Apply quality bonuses for 2026 standards."""
    if has_frontmatter:
        score += BONUS_HAS_FRONTMATTER
    if has_anchors:
        score += BONUS_HAS_ANCHORS
    return score


def score_content(markdown: str, links: list[str] | None = None) -> QualityScore:
    # Handle MagicMocks in tests
    if not isinstance(markdown, str):
        return QualityScore(1.0, False, False, False, False, True)

    text = (markdown or "").strip()
    links = links or []

    too_short = len(text) < THRESHOLD_MIN_CHARS
    missing_links = len(links) == 0
    duplicate_heavy = _check_duplicates(text)
    text_lower = text.lower()
    noisy = _compute_noise(text_lower)
    jargon_heavy = _check_jargon(text_lower)
    has_frontmatter = _check_frontmatter(text)
    has_anchors = _check_anchors(text)
    bot_challenge = is_bot_challenge(text)

    score = _compute_penalties(1.0, too_short, missing_links, duplicate_heavy, noisy, jargon_heavy)
    score = _compute_bonuses(score, has_frontmatter, has_anchors)
    score = max(0.0, min(1.0, score))
    acceptable = score >= ACCEPTABLE_THRESHOLD and not too_short and not bot_challenge

    return QualityScore(
        score=score,
        too_short=too_short,
        missing_links=missing_links,
        duplicate_heavy=duplicate_heavy,
        noisy=noisy or jargon_heavy,
        acceptable=acceptable,
        bot_challenge=bot_challenge,
    )
